File: webapp/go/src/isucon8/isucoin/model/trade.go
package model

import (
	"database/sql"
	"fmt"
	"isucon8/isubank"
	"log"
	"time"

	"github.com/pkg/errors"
)

//go:generate scanner
type Trade struct {
	ID        int64     `json:"id"`
	Amount    int64     `json:"amount"`
	Price     int64     `json:"price"`
	CreatedAt time.Time `json:"created_at"`
}

//go:generate scanner
type CandlestickData struct {
	Time  time.Time `json:"time"`
	Open  int64     `json:"open"`
	Close int64     `json:"close"`
	High  int64     `json:"high"`
	Low   int64     `json:"low"`
}

func GetTradeByID(d QueryExecutor, id int64) (*Trade, error) {
	return scanTrade(d.Query("SELECT * FROM trade WHERE id = ?", id))
}

func GetLatestTrade(d QueryExecutor) (*Trade, error) {
	return scanTrade(d.Query("SELECT * FROM trade ORDER BY id DESC"))
}

func GetCandlestickData(d QueryExecutor, mt time.Time, tf string) ([]*CandlestickData, error) {
	query := fmt.Sprintf(`
		SELECT m.t, a.price, b.price, m.h, m.l
		FROM (
			SELECT
				STR_TO_DATE(DATE_FORMAT(created_at, '%s'), '%s') AS t,
				MIN(id) AS min_id,
				MAX(id) AS max_id,
				MAX(price) AS h,
				MIN(price) AS l
			FROM trade
			WHERE created_at >= ?
			GROUP BY t
		) m
		JOIN trade a ON a.id = m.min_id
		JOIN trade b ON b.id = m.max_id
		ORDER BY m.t
	`, tf, "%Y-%m-%d %H:%i:%s")
	return scanCandlestickDatas(d.Query(query, mt))
}

func HasTradeChanceByOrder(d QueryExecutor, orderID int64) (bool, error) {
	order, err := GetOrderByID(d, orderID)
	if err != nil {
		return false, err
	}

	lowest, err := GetLowestSellOrder(d)
	switch {
	case err == sql.ErrNoRows:
		return false, nil
	case err != nil:
		return false, errors.Wrap(err, "GetLowestSellOrder")
	}

	highest, err := GetHighestBuyOrder(d)
	switch {
	case err == sql.ErrNoRows:
		return false, nil
	case err != nil:
		return false, errors.Wrap(err, "GetHighestBuyOrder")
	}

	switch order.Type {
	case OrderTypeBuy:
		if lowest.Price <= order.Price {
			return true, nil
		}
	case OrderTypeSell:
		if order.Price <= highest.Price {
			return true, nil
		}
	default:
		return false, errors.Errorf("other type [%s]", order.Type)
	}
	return false, nil
}

func reserveOrder(d QueryExecutor, order *Order, price int64) (int64, error) {
	bank, err := Isubank(d)
	if err != nil {
		return 0, errors.Wrap(err, "isubank init failed")
	}
	p := order.Amount * price
	if order.Type == OrderTypeBuy {
		p *= -1
	}

	id, err := bank.Reserve(order.User.BankID, p)
	if err != nil {
		if err == isubank.ErrCreditInsufficient {
			if derr := cancelOrder(d, order, "reserve_failed"); derr != nil {
				return 0, derr
			}
			sendLog(d, order.Type+".error", map[string]interface{}{
				"error":   err.Error(),
				"user_id": order.UserID,
				"amount":  order.Amount,
				"price":   price,
			})
			return 0, err
		}
		return 0, errors.Wrap(err, "isubank.Reserve")
	}

	return id, nil
}

func commitReservedOrder(tx *sql.Tx, order *Order, targets []*Order, reserves []int64) error {
	res, err := tx.Exec(`INSERT INTO trade (amount, price, created_at) VALUES (?, ?, NOW(6))`, order.Amount, order.Price)
	if err != nil {
		return errors.Wrap(err, "insert trade")
	}
	tradeID, err := res.LastInsertId()
	if err != nil {
		return errors.Wrap(err, "lastInsertID for trade")
	}
	sendLog(tx, "trade", map[string]interface{}{
		"trade_id": tradeID,
		"price":    order.Price,
		"amount":   order.Amount,
	})
	for _, o := range append(targets, order) {
		if _, err = tx.Exec(`UPDATE orders SET trade_id = ?, closed_at = NOW(6) WHERE id = ?`, tradeID, o.ID); err != nil {
			return errors.Wrap(err, "update order for trade")
		}
		sendLog(tx, o.Type+".trade", map[string]interface{}{
			"order_id": o.ID,
			"price":    order.Price,
			"amount":   o.Amount,
			"user_id":  o.UserID,
			"trade_id": tradeID,
		})
	}
	bank, err := Isubank(tx)
	if err != nil {
		return errors.Wrap(err, "isubank init failed")
	}
	if err = bank.Commit(reserves); err != nil {
		return errors.Wrap(err, "commit")
	}
	return nil
}

func tryTrade(tx *sql.Tx, orderID int64) error {
	order, err := getOpenOrderByID(tx, orderID)
	if err != nil {
		return err
	}

	restAmount := order.Amount
	unitPrice := order.Price
	reserves := make([]int64, 1, order.Amount+1)
	targets := make([]*Order, 0, order.Amount)

	reserves[0], err = reserveOrder(tx, order, unitPrice)
	if err != nil {
		return err
	}
	defer func() {
		if len(reserves) > 0 {
			bank, err := Isubank(tx)
			if err != nil {
				log.Printf("[WARN] isubank init failed. err:%s", err)
				return
			}
			if err = bank.Cancel(reserves); err != nil {
				log.Printf("[WARN] isubank cancel failed. err:%s", err)
			}
		}
	}()

	var targetOrders []*Order
	switch order.Type {
	case OrderTypeBuy:
		targetOrders, err = scanOrders(tx.Query(`SELECT * FROM orders WHERE type = ? AND closed_at IS NULL AND price <= ? ORDER BY price ASC, created_at ASC, id ASC`, OrderTypeSell, order.Price))
	case OrderTypeSell:
		targetOrders, err = scanOrders(tx.Query(`SELECT * FROM orders WHERE type = ? AND closed_at IS NULL AND price >= ? ORDER BY price DESC, created_at ASC, id ASC`, OrderTypeBuy, order.Price))
	}
	if err != nil {
		return errors.Wrap(err, "find target orders")
	}
	if len(targetOrders) == 0 {
		return ErrNoOrderForTrade
	}

	for _, to := range targetOrders {
		to, err = getOpenOrderByID(tx, to.ID)
		if err != nil {
			if err == ErrOrderAlreadyClosed {
				continue
			}
			return errors.Wrap(err, "getOpenOrderByID  buy_order")
		}
		if to.Amount > restAmount {
			continue
		}
		rid, err := reserveOrder(tx, to, unitPrice)
		if err != nil {
			if err == isubank.ErrCreditInsufficient {
				continue
			}
			return err
		}
		reserves = append(reserves, rid)
		targets = append(targets, to)
		restAmount -= to.Amount
		if restAmount == 0 {
			break
		}
	}
	if restAmount > 0 {
		return ErrNoOrderForTrade
	}
	if err = commitReservedOrder(tx, order, targets, reserves); err != nil {
		return err
	}
	reserves = reserves[:0]
	return nil
}

func RunTrade(db *sql.DB) error {
	lowestSellOrder, err := GetLowestSellOrder(db)
	switch {
	case err == sql.ErrNoRows:
		// 売り注文が無いため成立しない
		return nil
	case err != nil:
		return errors.Wrap(err, "GetLowestSellOrder")
	}

	highestBuyOrder, err := GetHighestBuyOrder(db)
	switch {
	case err == sql.ErrNoRows:
		// 買い注文が無いため成立しない
		return nil
	case err != nil:
		return errors.Wrap(err, "GetHighestBuyOrder")
	}

	if lowestSellOrder.Price > highestBuyOrder.Price {
		// 最安の売値が最高の買値よりも高いため成立しない
		return nil
	}

	candidates := make([]int64, 0, 2)
	if lowestSellOrder.Amount > highestBuyOrder.Amount {
		candidates = append(candidates, lowestSellOrder.ID, highestBuyOrder.ID)
	} else {
		candidates = append(candidates, highestBuyOrder.ID, lowestSellOrder.ID)
	}

	for _, orderID := range candidates {
		err := func() error {
			tx, err := db.Begin()
			if err != nil {
				return errors.Wrap(err, "begin transaction failed")
			}
			err = tryTrade(tx, orderID)
			switch err {
			case nil, ErrNoOrderForTrade, ErrOrderAlreadyClosed, isubank.ErrCreditInsufficient:
				tx.Commit()
			default:
				tx.Rollback()
			}
			return err
		}()
		switch err {
		case nil:
			// トレード成立したため次の取引を行う
			return RunTrade(db)
		case ErrNoOrderForTrade, ErrOrderAlreadyClosed:
			// 注文個数の多い方で成立しなかったので少ない方で試す
			continue
		default:
			return err
		}
	}
	// 個数のが不足していて不成立
	return nil
}

File: webapp/php/src/Isucon8/Isucoin/Model/Trade.php
<?php

use Isucon8\PDOWrapper;
use Isucon8\PDOWrapperTxn;
use Isucon8\Isubank\CreditInsufficientException as IsubankCreditInsufficientException;

class JsonableTrade extends ArrayObject implements JsonSerializable {
    public function jsonSerialize() {
        return [
            'id'         => (int)$this['id'],
            'amount'     => (int)$this['amount'],
            'price'      => (int)$this['price'],
            'created_at' => (string)(new DateTime($this['created_at']))->format('Y-m-d\TH:i:sP'),
        ];
    }
}

class JsonableCandlestickData extends ArrayObject implements JsonSerializable {
    public function jsonSerialize() {
        return [
            'time'  => (string)(new DateTime($this['t']))->format('Y-m-d\TH:i:sP'),
            'open'  => (int)$this['open'],
            'close' => (int)$this['close'],
            'high'  => (int)$this['h'],
            'low'   => (int)$this['l'],
        ];
    }
}

function reformTrade(array $trade): JsonableTrade {
    return new JsonableTrade($trade);
}

function GetTradeByID(PDOWrapper $dbh, int $id): JsonableTrade {
    return reformTrade($dbh->select_row('SELECT * FROM trade WHERE id = ?', $id));
}

function GetLatestTrade(PDOWrapper $dbh): JsonableTrade {
    return reformTrade($dbh->select_row('SELECT * FROM trade ORDER BY id DESC'));
}

function GetCandlestickData(PDOWrapper $dbh, DateTime $mt, string $tf): array {
    $sql = <<< SQL
SELECT m.t, a.price AS open, b.price AS close, m.h, m.l
FROM (
    SELECT
        STR_TO_DATE(DATE_FORMAT(created_at, '%s'), '%s') AS t,
        MIN(id) AS min_id,
        MAX(id) AS max_id,
        MAX(price) AS h,
        MIN(price) AS l
    FROM trade
    WHERE created_at >= ?
    GROUP BY t
) m
JOIN trade a ON a.id = m.min_id
JOIN trade b ON b.id = m.max_id
ORDER BY m.t
SQL;

    $query = sprintf($sql, $tf, '%Y-%m-%d %H:%i:%s');

    $rows = [];
    foreach ($dbh->select_all($query, $mt->format('Y-m-d H:i:s')) as $row) {
        $rows[] = new JsonableCandlestickData($row);
    }
    return $rows;
}

function HasTradeChanceByOrder(PDOWrapper $dbh, int $order_id): bool {
    $order = null;
    try {
        $order = GetOrderByID($dbh, $order_id);
    } catch(\Throwable $throwable) {
        throw $throwable;
    }

    $lowest = null;
    try {
        $lowest = GetLowestSellOrder($dbh);
    } catch(Isucon8\NoRowsException $e) {
        return false;
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: GetLowestSellOrder failed', $throwable->getMessage()));
        throw $throwable;
    }

    $highest = null;
    try {
        $highest = GetHighestBuyOrder($dbh);
    } catch(Isucon8\NoRowsException $e) {
        return false;
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: GetHighestBuyOrder failed', $throwable->getMessage()));
        throw $throwable;
    }


    switch ($order['type']) {
    case ORDER_TYPE_BUY:
        if ($lowest['price'] <= $order['price']) {
            return true;
        }
        break;
    case ORDER_TYPE_SELL:
        if ($order['price'] <= $highest['price']) {
            return true;
        }
        break;
    default:
        throw new Exception('other type [%s]', $order['type']);
    }
    return false;
}

function reserveOrder(PDOWrapper $dbh, JsonableOrder $order, int $price): int {
    $bank = null;
    try {
        $bank = Isubank($dbh);
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: isubank init failed', $throwable->getMessage()));
        throw $throwable;
    }

    $p = $order['amount'] * $price;
    if ($order['type'] === ORDER_TYPE_BUY) {
        $p *= -1;
    }

    $id = null;
    try {
        $id = $bank->reserve($order['user']['bank_id'], $p);
    } catch(IsubankCreditInsufficientException $e) {
        try {
            cancelOrder($dbh, $order, 'reserve_failed');
        } catch(\Throwable $throwable) {
            error_log(sprintf('%s: cancelOrder failed', $throwable->getMessage()));
            throw $throwable;
        }
        sendLog($dbh, $order['type'].'.error', [
            'error' => (string)$e->getMessage(),
            'user_id' => (int)$order['user_id'],
            'amount' => (int)$order['amount'],
            'price' => (int)$price,
        ]);
        throw $e;
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: isubank.Reserve', $throwable->getMessage()));
        throw $throwable;
    }

    return $id;
}

function commitReservedOrder(PDOWrapperTxn $tx, JsonableOrder $order, array $targets, array $reserves): void {
    try {
        $tx->execute('INSERT INTO trade (amount, price, created_at) VALUES (?, ?, NOW(6))', $order['amount'], $order['price']);
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: insert trade', $throwable->getMessage()));
        throw $throwable;
    }

    $trade_id = $tx->last_insert_id();

    sendLog($tx, 'trade', [
        'trade_id' => (int)$trade_id,
        'price' => (int)$order['price'],
        'amount' => (int)$order['amount'],
    ]);

    $targets[] = $order;

    foreach ($targets as $o) {
        try {
            $tx->execute('UPDATE orders SET trade_id = ?, closed_at = NOW(6) WHERE id = ?', $trade_id, $o['id']);
        } catch(\Throwable $throwable) {
            error_log(sprintf('%s: update order for trade', $throwable->getMessage()));
            throw $throwable;
        }
        sendLog($tx, $o['type'].'.trade', [
            'order_id' => (int)$o['id'],
            'price' => (int)$order['price'],
            'amount' => (int)$o['amount'],
            'user_id' => (int)$o['user_id'],
            'trade_id' => (int)$trade_id,
        ]);
    }

    $bank = null;
    try {
        $bank = Isubank($tx);
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: isubank init failed', $throwable->getMessage()));
        throw $throwable;
    }

    try {
        $bank->commit($reserves);
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: commit', $throwable->getMessage()));
        throw $throwable;
    }
}

function tryTrade(PDOWrapperTxn $tx, int $order_id): void {
    $order = null;
    try {
        $order = getOpenOrderByID($tx, $order_id);
    } catch(\Throwable $throwable) {
        throw $throwable;
    }

    $rest_amount = $order['amount'];
    $unit_price = $order['price'];
    $reserves = [0];
    $targets = [];

    try {
        $reserves[0] = reserveOrder($tx, $order, $unit_price);

        $target_orders = [];
        try {
            switch ($order['type']) {
            case ORDER_TYPE_BUY:
                $target_orders = queryOrders($tx, 'SELECT * FROM orders WHERE type = ? AND closed_at IS NULL AND price <= ? ORDER BY price ASC, created_at ASC, id ASC', ORDER_TYPE_SELL, $order['price']);
                break;
            case ORDER_TYPE_SELL:
                $target_orders = queryOrders($tx, 'SELECT * FROM orders WHERE type = ? AND closed_at IS NULL AND price >= ? ORDER BY price DESC, created_at ASC, id ASC', ORDER_TYPE_BUY, $order['price']);
                break;
            }
        } catch(\Throwable $throwable) {
            error_log(sprintf('%s: find target orders', $throwable->getMessage()));
            throw $throwable;
        }
        if (count($target_orders) === 0) {
            throw new NoOrderForTradeException();
        }

        foreach ($target_orders as $to) {
            try {
                $to = getOpenOrderByID($tx, $to['id']);
            } catch(OrderAlreadyClosedException $e) {
                continue;
            } catch(\Throwable $throwable) {
                error_log(sprintf('%s: getOpenOrderByID buy_order', $throwable->getMessage()));
                throw $throwable;
            }

            if ($to['amount'] > $rest_amount) {
                continue;
            }

            $rid = null;
            try {
                $rid = reserveOrder($tx, $to, $unit_price);
            } catch(IsubankCreditInsufficientException $e) {
                continue;
            } catch(\Throwable $throwable) {
                throw $throwable;
            }
            $reserves[] = $rid;
            $targets[] = $to;
            $rest_amount -= $to['amount'];
            if ($rest_amount === 0) {
                break;
            }
        }

        if (0 < $rest_amount) {
            throw new NoOrderForTradeException();
        }

        try {
            commitReservedOrder($tx, $order, $targets, $reserves);
        } catch(\Throwable $throwable) {
            throw $throwable;
        }
        $reserves = [];
    } finally {
        if (0 < count($reserves)) {
            $bank = null;
            try {
                $bank = Isubank($tx);
            } catch(\Throwable $throwable) {
                error_log(sprintf('[WARN] isubank init failed. err:%s', $throwable->getMessage()));
            }
            if ($bank !== null) {
                try {
                    $bank->cancel($reserves);
                } catch(\Throwable $throwable) {
                    error_log(sprintf('[WARN] isubank cancel failed. err:%s', $throwable->getMessage()));
                }
            }
        }
    }
}

function RunTrade(PDOWrapper $dbh): void {
    $lowest_sell_order = null;
    try {
        $lowest_sell_order = GetLowestSellOrder($dbh);
    } catch(Isucon8\NoRowsException $e) {
        // 売り注文が無いため成立しない
        return;
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: GetLowestSellOrder failed', $throwable->getMessage()));
        throw $throwable;
    }

    $highest_buy_order = null;
    try {
        $highest_buy_order = GetHighestBuyOrder($dbh);
    } catch(Isucon8\NoRowsException $e) {
        // 買い注文が無いため成立しない
        return;
    } catch(\Throwable $throwable) {
        error_log(sprintf('%s: GetHighestBuyOrder failed', $throwable->getMessage()));
        throw $throwable;
    }

    if ( $lowest_sell_order['price'] > $highest_buy_order['price']) {
        // 最安の売値が最高の買値よりも高いため成立しない
        return;
    }

    $candidates = null;
    if ($lowest_sell_order['amount'] > $highest_buy_order['amount']) {
        $candidates = [$lowest_sell_order['id'], $highest_buy_order['id']];
    } else {
        $candidates = [$highest_buy_order['id'], $lowest_sell_order['id']];
    }

    foreach ($candidates as $order_id) {
        try {
            $tx = null;
            try {
                $tx = $dbh->beginTxn();
            } catch(\Throwable $throwable) {
                error_log(sprintf('%s: begin transaction failed', $throwable->getMessage()));
                throw $throwable;
            }

            try {
                tryTrade($tx, $order_id);
                $tx->commit();
            } catch(NoOrderForTradeException | OrderAlreadyClosedException | IsubankCreditInsufficientException $e) {
                $tx->commit();
                throw $e;
            } catch(\Throwable $throwable) {
                $tx->rollback();
                throw $throwable;
            }
        } catch(NoOrderForTradeException | OrderAlreadyClosedException $e) {
            // 注文個数の多い方で成立しなかったので少ない方で試す
            continue;
        } catch(\Throwable $throwable) {
            throw $throwable;
        }
        // トレード成立したため次の取引を行う
        RunTrade($dbh);
    }
    // 個数のが不足していて不成立
    return;
}
